### ISSUE TITLE
AddressSanitizer error in `CGO_getFakeDataChannel` when `-fsanitize=address` is added to `CGO_CXXFLAGS`

### ISSUE THREAD

**Issue by uumaro** [keroserene/go-webrtc#95]
Pull request #94 shows that adding `-fsanitize=address` to `CGO_CXXFLAGS` reveals an error:

```
$ CGO_CFLAGS="-fsanitize=address" CGO_CXXFLAGS="-fsanitize=address" CGO_LDFLAGS="-fsanitize=address -fuse-ld=gold" go test -run TestDataStateEnums
  DataChannel ✔✔✔✔✔✔✔✔✔✔
    Callbacks should fire correctly 
      OnMessage ✔✔✔✔✔✔✔✔✔✔✔✔
      StateChangeCallbacks ✔✔✔✔✔✔=================================================================
==5778==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7ffe9de01468 at pc 0x0000005b7aaa bp 0x7ffe9de01420 sp 0x7ffe9de01418
WRITE of size 8 at 0x7ffe9de01468 thread T0
    #0 0x5b7aa9 in rtc::scoped_refptr<webrtc::DataChannelInterface>::scoped_refptr<FakeDataChannel>(rtc::scoped_refptr<FakeDataChannel> const&) (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5b7aa9)
    #1 0x5b74f0 in rtc::RefCountedObject<CGoDataChannelObserver>::RefCountedObject<rtc::scoped_refptr<FakeDataChannel>&>(rtc::scoped_refptr<FakeDataChannel>&) (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5b74f0)
    #2 0x5ad307 in CGO_getFakeDataChannel (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5ad307)
    #3 0x5a909a in _cgo_aeba63b75ef5_Cfunc_CGO_getFakeDataChannel (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5a909a)
    #4 0x49c2ef in runtime.asmcgocall (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x49c2ef)
Address 0x7ffe9de01468 is located in stack of thread T0==5778==AddressSanitizer CHECK failed: ../../../../src/libsanitizer/asan/asan_thread.cc:231 "((ptr[0] == kCurrentStackFrameMagic)) != (0)" (0x0, 0x0)
    #0 0x7f7553c0fff4  (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x9bff4)
    #1 0x7f7553c14a73 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/usr/lib/x86_64-linux-gnu/libasan.so.2+0xa0a73)
    #2 0x7f7553c127b9  (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x9e7b9)
    #3 0x7f7553c0d1f9  (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x991f9)
    #4 0x7f7553c0d7da  (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x997da)
    #5 0x7f7553c0f006 in __asan_report_error (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x9b006)
    #6 0x7f7553c10646 in __asan_report_store8 (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x9c646)
    #7 0x5b7aa9 in rtc::scoped_refptr<webrtc::DataChannelInterface>::scoped_refptr<FakeDataChannel>(rtc::scoped_refptr<FakeDataChannel> const&) (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5b7aa9)
    #8 0x5b74f0 in rtc::RefCountedObject<CGoDataChannelObserver>::RefCountedObject<rtc::scoped_refptr<FakeDataChannel>&>(rtc::scoped_refptr<FakeDataChannel>&) (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5b74f0)
    #9 0x5ad307 in CGO_getFakeDataChannel (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5ad307)
    #10 0x5a909a in _cgo_aeba63b75ef5_Cfunc_CGO_getFakeDataChannel (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x5a909a)
    #11 0x49c2ef in runtime.asmcgocall (/tmp/go-build464808380/github.com/keroserene/go-webrtc/_test/go-webrtc.test+0x49c2ef)
exit status 1
FAIL	github.com/keroserene/go-webrtc	0.045s
The command "if [[ "$OS" == "linux" ]]; then CGO_CFLAGS="-fsanitize=address" CGO_CXXFLAGS="-fsanitize=address" CGO_LDFLAGS="-fsanitize=address -fuse-ld=gold" go test -v .; fi" exited with 1.
```

**Comment by uumaro** [comment]
I get the error both in the Travis CI environment (Ubuntu 16.04 xenial) and on my desktop with Debian 9 (stretch). But interestingly, I don't get the error, nor any other AddressSanitizer error, on a laptop with Debian 10 (buster).

Versions that result in the error:

| package | version |
|-----|---------|
| gcc/libasan3 | 6.3.0-18+deb9u1 |
| go           | 1.7.4           |

Versions that do not result in the error:

| package | version |
|-----|---------|
| gcc/libasan5 | 8.2.0-8         |
| go           | 1.10.4          |

**Comment by uumaro** [comment]
I spent some time looking at this function and understanding [`rtc::scoped_refptr`](https://chromium.googlesource.com/external/webrtc/stable/src/+/ba74f4d931c4b968cfdf87db7ffe52e20435657c/system_wrappers/interface/scoped_refptr.h#13), and I don't think there's a problem. Certainly not a stack underflow.
```
rtc::scoped_refptr<CGoDataChannelObserver> test_observer;

void* CGO_getFakeDataChannel() {
  rtc::scoped_refptr<FakeDataChannel> test_dc = new rtc::RefCountedObject<FakeDataChannel>();
  test_observer = new rtc::RefCountedObject<CGoDataChannelObserver>(test_dc);
  auto o = test_observer.get();
  test_dc->RegisterObserver(o);
  return (void *)o;
}
```
`test_dc` only lives for the body of the function, but `test_observer` gets a copy with its own independent reference count in the constructor of `CGoDataChannelObserver`, and `test_observer` is global so it outlives the function call. `test_observer` owns the memory of `o`, so `o` is alive after being returned to the caller.

The only weird thing I can spot is that `CGO_getFakeDataChannel` overwrites the global `test_observer`--when you call it a second time the first value goes out of scope and is deleted. But as things stand, `CGO_getFakeDataChannel` is not called with overlapping lifetimes, so that problem wouldn't account for any errors we're seeing.

----
I compiled a test program that only calls `CGO_getFakeDataChannel` (i.e., C++ code only, no Cgo) and ran it with valgrind, and didn't find any problems. Here is the program:
```
#include <iostream>

#include "datachannel.h"
#include "datachannel.hpp"

int main()
{
  CGoDataChannelObserver *p;

  p = (CGoDataChannelObserver *) CGO_getFakeDataChannel();
  std::cout << "p " << p << std::endl;
  std::cout << "p->dc " << p->dc << std::endl;

  return 0;
}
```
To make it compile, I had to comment out a few lines elsewhere:
```diff
diff --git a/datachannel.hpp b/datachannel.hpp
index 4a52e9b..55033a0 100644
--- a/datachannel.hpp
+++ b/datachannel.hpp
@@ -1,7 +1,7 @@
 #ifndef _DATACHANNEL_H_
 #define _DATACHANNEL_H_

-#include <_cgo_export.h>  // Allow calling certain Go functions.
+// #include <_cgo_export.h>  // Allow calling certain Go functions.
 #include <assert.h>

 #include "api/peerconnectioninterface.h"
@@ -18,17 +18,17 @@ class CGoDataChannelObserver
   }

   void OnStateChange() {
-    cgoChannelOnStateChange(goChannel);
+    // cgoChannelOnStateChange(goChannel);
   }

   void OnMessage(const webrtc::DataBuffer& buffer) {
     auto data = reinterpret_cast<void*>(const_cast<unsigned char*>(
                                         buffer.data.data()));
-    cgoChannelOnMessage(goChannel, data, buffer.size());
+    // cgoChannelOnMessage(goChannel, data, buffer.size());
   }

   void OnBufferedAmountChange(uint64_t previous_amount) {
-    cgoChannelOnBufferedAmountChange(goChannel, previous_amount);
+    // cgoChannelOnBufferedAmountChange(goChannel, previous_amount);
   }

   // Reference to external Go data.Channel required for callbacks.
```
I compiled and ran it like this:
```
$ g++ -g -I include/ -L lib/ -o test test.cc datachannel.cc -l webrtc-linux-amd64-magic -lpthread
$ valgrind ./test 
==30861== Command: ./test
==30861== 
p 0x5ca2ce0
p->dc 0x5ca2c80
==30861== 
```

----

On the other hand, as I noted above, if you call `CGO_getFakeDataChannel` twice, then the second call overwrites the global variable `test_observer` and you may be left with dangling references.
```
#include <iostream>

#include "datachannel.h"
#include "datachannel.hpp"

int main()
{
  CGoDataChannelObserver *p1, *p2;

  p1 = (CGoDataChannelObserver *) CGO_getFakeDataChannel();
  std::cout << "p1 " << p1 << std::endl;
  std::cout << "p1->dc " << p1->dc << std::endl;

  std::cout << std::endl;

  p2 = (CGoDataChannelObserver *) CGO_getFakeDataChannel();
  std::cout << "p2 " << p2 << std::endl;
  std::cout << "p2->dc " << p2->dc << std::endl;

  std::cout << std::endl;

  std::cout << "p1 " << p1 << std::endl;
  // use-after-free occurs here
  std::cout << "p1->dc " << p1->dc << std::endl;

  return 0;
}
```
And valgrind picks up the error:
```
==29550== Command: ./test2
==29550==
p1 0x5ca2ce0
p1->dc 0x5ca2c80

p2 0x5ca31f0
p2->dc 0x5ca3190

p1 0x5ca2ce0
==29550== Invalid read of size 8
==29550==    at 0x10E48A: rtc::scoped_refptr<webrtc::DataChannelInterface>::operator webrtc::DataChannelInterface*() const (scoped_ref_ptr.h:103)
==29550==    by 0x10CF80: main (test.cc:25)
==29550==  Address 0x5ca2cf8 is 24 bytes inside a block of size 40 free'd
==29550==    at 0x4C2D2DB: operator delete(void*) (vg_replace_malloc.c:576)
==29550==    by 0x11156A: rtc::RefCountedObject<CGoDataChannelObserver>::~RefCountedObject() (refcountedobject.h:54)
==29550==    by 0x1115DC: rtc::RefCountedObject<CGoDataChannelObserver>::Release() const (refcountedobject.h:40)
==29550==    by 0x11113A: rtc::scoped_refptr<CGoDataChannelObserver>::operator=(CGoDataChannelObserver*) (scoped_ref_ptr.h:122)
==29550==    by 0x10EF11: CGO_getFakeDataChannel (datachannel.cc:162)
==29550==    by 0x10CE9C: main (test.cc:17)
==29550==  Block was alloc'd at
==29550==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==29550==    by 0x10EEF0: CGO_getFakeDataChannel (datachannel.cc:162)
==29550==    by 0x10CDFD: main (test.cc:11)
==29550==
p1->dc 0x5ca2c80
==29550==
```

**Comment by arlolra** [comment]
Putting 94aa8d8 here to think about.

https://travis-ci.org/keroserene/go-webrtc/jobs/462360802

**Comment by arlolra** [comment]
Did you try compiling the test (`go test -c`) and looking at the underflow in a debugger?

I'll need to setup a vm where this is reproducible.

**Comment by uumaro** [comment]
> Did you try compiling the test (`go test -c`) and looking at the underflow in a debugger?

That was a good suggestion. I tried that, and found that my [earlier belief](https://github.com/keroserene/go-webrtc/issues/95#issuecomment-443399524) that `CGO_getFakeDataChannel` is called only once is wrong. In fact it gets called 6 times during the tests, which means that the global `test_observer` is being overwritten. Not every call results in an error; in my test it would happen on the third call. I would have expected some use-after-free error rather than the stack underflow. It happens even when I use `-test.parallel 1`.

I tried your https://github.com/keroserene/go-webrtc/commit/94aa8d8d588d1aa32368ad84fd85547647420d65 and added a debugging output to print the size of `test_observers`.
```
Now 1 observers
Now 2 observers
Now 3 observers
Now 4 observers
Now 5 observers
Now 6 observers
```

I think I was misunderstanding how GoConvey works. It seems that the overwriting of the global variable that I assumed didn't affect us, does.

**Comment by arlolra** [comment]
Looking at an example from the GoConvey docs,
https://github.com/smartystreets/goconvey/blob/master/convey/isolated_execution_test.go#L341-L375

It seems like `CGO_getFakeDataChannel` would be called for each leaf in `Convey("DataChannel", t, func() { ... })`, of which there are six, so that checks out.

But here's something to think about.  If you comment out,

```
So(func() {
	cgoFakeStateChange(c, 999)
}, ShouldPanic)
```

You no longer get the underflow, even with the dangling pointers.  That test expects a panic called through cgo, which [according to the issue filed at go](https://github.com/golang/go/issues/16150#issuecomment-256918423), isn't expected to work.

Mind you, the stack at that link is pretty much the trace we're looking at here, so I'm wondering how circular this all is.

On that note, if you delete the other conflicting tests and only run `Convey("StateChangeCallbacks", func() { ... })` with the panic check still in, there isn't any problem.

**Comment by uumaro** [comment]
> But here's something to think about. If you comment out,
> 
> ```
> So(func() {
> 	cgoFakeStateChange(c, 999)
> }, ShouldPanic)
> ```
> 
> You no longer get the underflow, even with the dangling pointers. That test expects a panic called through cgo, which [according to the issue filed at go](https://github.com/golang/go/issues/16150#issuecomment-256918423), isn't expected to work.

I think you're on to something here, with the `panic` being a non-local jump and (presumably) messing up AddressSanitizer's view of the stack somewhere.

> On that note, if you delete the other conflicting tests and only run `Convey("StateChangeCallbacks", func() { ... })` with the panic check still in, there isn't any problem.

It doesn't seem to be the `panic` itself that causes the AddressSanitizer error, but the next call to `CGO_getFakeDataChannel` after the `panic`. If you comment out all the tests except `"StateChangeCallbacks"` (which panics) and `"OnBufferedAmountLow"`, you get the AddressSanitizer error in the second.

In any event, I think your idea in https://github.com/keroserene/go-webrtc/commit/94aa8d8d588d1aa32368ad84fd85547647420d65 of keeping live references to all the past test observers is a good one. It would be nicer if they were instead owned by the caller instead of a global, but it looks like that's not trivial, and it's okay for testing purposes.

**Comment by uumaro** [comment]
> > In any event, I think your idea in [94aa8d8](https://github.com/keroserene/go-webrtc/commit/94aa8d8d588d1aa32368ad84fd85547647420d65) of keeping live references to all the past test observers is a good one. It would be nicer if they were instead owned by the caller instead of a global, but it looks like that's not trivial, and it's okay for testing purposes.
> 
> Keeping that memory alive is fine, but probably is just masking the issue. What we really need to do is comment out the panic until the upstream issue resolved.

That's okay with me. We need to either drop the panic test or drop AddressSanitizer. (Or both, if there are more problems beside the panic.) I'll comment on #100.
